server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java

```diff
@@ -113,30 +113,41 @@ public static LayerAndLoader ofLoader(ClassLoader loader) {
     private final Map<String, LoadedPluginLayer> loadedPluginLayers;
 
     /**
-     * Constructs a new PluginsLoader
+     * Constructs a new PluginsLoader. With this factory method, the set of qualified exports/opens to Java modular modules/plugins will
+     * include all qualified exports from the server module. This is the desired behaviour in production use.
      *
      * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
      * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
      */
     public static PluginsLoader createPluginsLoader(Path modulesDirectory, Path pluginsDirectory) {
-        return createPluginsLoader(modulesDirectory, pluginsDirectory, true);
+        Map<String, List<ModuleQualifiedExportsService>> qualifiedExports = new HashMap<>(ModuleQualifiedExportsService.getBootServices());
+        addServerExportsService(qualifiedExports);
+        return createPluginsLoader(modulesDirectory, pluginsDirectory, qualifiedExports);
+    }
+
+    /**
+     * Constructs a new PluginsLoader. With this factory method, the set of qualified exports/opens to Java modular modules/plugins will
+     * not include qualified exports from the server module. This is the desired behaviour in test use, as tests are running non-modular.
+     *
+     * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
+     * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
+     */
+    public static PluginsLoader createPluginsLoaderWithoutServerExports(Path modulesDirectory, Path pluginsDirectory) {
+        return createPluginsLoader(modulesDirectory, pluginsDirectory, Map.copyOf(ModuleQualifiedExportsService.getBootServices()));
     }
 
     /**
      * Constructs a new PluginsLoader
      *
      * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
      * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
-     * @param withServerExports {@code true} to add server module exports
+     * @param qualifiedExports The set of qualified exports/opens to use when constructing Java modular modules/plugins
```

Comment:
This definition doesn't seem right. These are not equal exports/opens. Rather, it is a hole into a module that allows extending that module dynamically. It's specifically needed for modules that are dependencies of server which have qualified exports, so that the module being exported to can regain access after it has been created.

The only way to add exports/opens to a module after it is loaded is by the module modifying itself. A ModuleQualifiedExportService is a tool that makes this process easier. A subclass is created and only 2 methods must be implemented, all with the same implementation, but that implementation must live within the module itself so that the call to modify the module lives inside the module.

All that is to say, these aren't actually exports, and I think this confusion may be making the plugins loader more complicated.